### PR TITLE
Editable installs

### PR DESCRIPTION
**Pull request by gaborbernat**
Things Paul would like to see changed/amended:

- mapping of paths for purelib/platlib - reorganize source layout into a destination layout
- version number within the returned object
- better name for orchestrator; is it just frontend - but does that differentiates it from the build frontend - do we need to differentiate?

### CODE REVIEW

File: pep-9999.rst [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+(as introduced by `PEP-517`_) to allow projects to be installed in editable
+mode.
+
+Motivation
+==========
+
+Python programmers want to be able to develop packages without having to
+install them after every change. Installing a project is made up of
+building a wheel from the source directory (as described by `PEP-517`_
+and `PEP-518`_) and then install that wheel into a python environment.
```
Comment (pradyunsg) [inline pep-9999.rst:26]:
```suggestion
and `PEP-518`_) and then installing that wheel into a python environment.
```

Comment (jugmac00) [inline pep-9999.rst:26]:
```suggestion
and `PEP-518`_) and then installing that wheel into a Python environment.
```

File: pep-9999.rst [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+Currently users can achieve this by:
+
+- Adding the relevant source directories to ``sys.path`` (configurable from the
+  command line interface via the ``PYTHONPATH`` environment variable). Note in
+  this case the user have to install the project dependencies themselves and
```
Comment (jugmac00) [inline pep-9999.rst:32]:
```suggestion
  this case the users have to install the project dependencies themselves and
```

File: pep-9999.rst [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+  entry points or project metadata are not generated.
+- `setuptools`_ provides the `setup.py develop`_ mechanism: installs a
+  ``pth`` file that injects the project root onto the ``sys.path``
+  at interpreter startup time, generates the project metadata and also installs
+  project dependencies. `pip`_ exposes calling this mechanism via the
+  `pip install -e <project_directory>`_ command line interface.
+- `flit`_ provides the `flit install --symlink`_ command that: symlinks the
```
Comment (jugmac00) [inline pep-9999.rst:39]:
```suggestion
- `flit`_ provides the `flit install --symlink`_ command that symlinks the
```

File: pep-9999.rst [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+  project files into the interpreters ``purelib`` folder, generates the project
+  metadata and also installs dependencies.
+
+The installation of a project in a way that the code being imported and
+executed remains in the source directory is known as the *editable*
+installation mode. This mode allows the user to change the code and run it
+without needing to reinstall it after every change.
+
+The goal of this PEP is to provide a standardized interface between the build
+frontend and backends for building and installing projects in editable mode.
```
Comment (jugmac00) [inline pep-9999.rst:49]: Why is it "build frontend" but "build backends"? singular vs plural

File: pep-9999.rst [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+This will allow backends to expose the project in editable mode, and for the
+frontends a way to request this, in a generic mode for all build backend and
+frontends. The PEP also proposes on not imposing any editable mode onto the
```
Comment (jugmac00) [inline pep-9999.rst:52]: Again `backend` vs `frontends`.

File: pep-9999.rst [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+packaging tool the user communicates with (and in turn which uses the build
+frontend to communicate with the build backend).
```
Comment (jugmac00) [inline pep-9999.rst:54]:
```suggestion
packaging tool the user communicates with. In turn the packing tool uses the build
frontend to communicate with the build backend).
```

File: pep-9999.rst [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+Rationale
+=========
+
+`PEP-517`_ deferred "Editable installs" because this would have delayed further
+its adoption and there wasn't an agreement how editable installs should be
+achieved. Due to the popularity of the `setuptools`_ and `pip`_ projects
+the status quo prevailed, and backends could achieve editable mode by providing
+a ``setup.py develop`` implementation, which the user could trigger via
+`pip install -e <project_directory>`_. By defining an editable interface
+between build backend and frontends we can totally eliminate the need of the
```
Comment (jugmac00) [inline pep-9999.rst:65]: Again: `backend` vs `frontends`.

File: pep-9999.rst [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+``setup.py`` file and communication method.
```
Comment (jugmac00) [inline pep-9999.rst:66]: What is `and communication method.` referring to?

If this is referring to the **current** communication method between build frontend and backend, maybe rephrase it to 

```suggestion
``setup.py`` file and their current communication method.
```

File: pep-9999.rst [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+
+Once the project is installed in editable mode, users expect that changes to
+the project code in the local source tree become effective without the need of
+a new installation step.
+
+Some kind of changes, such as the addition or modification of entry points, or
+the addition of new dependencies, require a new installation step to become
+effective. These changes are typically made in build backend configuration
+files (such as ``pyproject.toml``), so it is consistent with the general user
+expectation that source code is imported from the source tree.
```
Comment (jugmac00) [inline pep-9999.rst:82]: I am not sure why there is the second part of the sentence about the user expectation and how this relates to the first part of the sentence with configuration changes. -> ??? for me :-) What is `it` referring to? Why are changes in configuration files consistent with how code is imported? I do not understand the intention of this sentences.

File: pep-9999.rst [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+
+When a project is installed in editable mode, users expect the installation to
+behave identically as a regular installation. Depending on the way build build
```
Comment (jugmac00) [inline pep-9999.rst:85]:
```suggestion
behave identically as a regular installation. Depending on the way build
```

File: pep-9999.rst [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+frontends implement this specification, some minor differences may be visible,
+such as the presence of additional files (compared to a regular installation)
```
Comment (jugmac00) [inline pep-9999.rst:87]:
```suggestion
such as the presence of additional files (compared to a regular installation),
```

File: pep-9999.rst [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+either in the source tree or the interpreters installation path. Both the build
```
Comment (jugmac00) [inline pep-9999.rst:88]:
```suggestion
either in the source tree or the interpreter's installation path. Both the build
```

File: pep-9999.rst [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+backend and frontends are encouraged to document such potential differences.
```
Comment (jugmac00) [inline pep-9999.rst:89]: Again: `backend` vs `frontends`

File: pep-9999.rst [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+
+For reference a non editable installation works as follows:
```
Comment (jugmac00) [inline pep-9999.rst:91]:
```suggestion
For reference, a non editable installation works as follows:
```

File: pep-9999.rst [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+
+1. The **developer** is using a tool, we'll call it here the **orchestrator**
+   to drive the project development (e.g., `pip`_). When the user wants to
+   trigger a package build and install of a project they'll communicate with
```
Comment (jugmac00) [inline pep-9999.rst:95]:
```suggestion
   trigger a package build and installation of a project they'll communicate with
```

File: pep-9999.rst [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+   the **orchestrator**.
+2. The orchestrator uses a **build frontend** to trigger build of a wheel
```
Comment (jugmac00) [inline pep-9999.rst:97]:
```suggestion
2. The orchestrator uses a **build frontend** to trigger the build of a wheel
```

File: pep-9999.rst [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+   (e.g., `build`_). The build frontend uses `PEP-517`_ to communicate with the
+   **build backend** (e.g. `setuptools`_) - with the build backend installed
+   into an isolated environment. Once invoked the backend returns a wheel.
+3. The orchestrator takes the wheel and feeds it to an **installer** (e.g.,
+   `installer`_) to install the wheel into the target python interpreter.
```
Comment (jugmac00) [inline pep-9999.rst:102]:
```suggestion
   `installer`_) to install the wheel into the target Python interpreter.
```

File: pep-9999.rst [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+the hooks is used to specify the build dependencies of an editable install, and
+the other hook returns the necessary information via the build frontend the
+orchestrator needs to create an editable install.
+
+get_requires_for_build_editable
+-------------------------------
+
+:: 
+
+  def build_editable(config_settings=None):
```
Comment (jugmac00) [inline pep-9999.rst:117]: Looks like this is the wrong function name here?

File: pep-9999.rst [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+      ...
+
+This hook MUST return an additional list of strings containing `PEP-508`_
+dependency specifications, above and beyond those specified in the
+``pyproject.toml`` file. These dependencies will be installed when calling the
+``build_editable`` hook.
```
Comment (jugmac00) [inline pep-9999.rst:123]: Just for my personal interest - which dependencies could this be?

Why does an editable install need more/different dependencies than a regular one?

File: pep-9999.rst [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+
+::
+
+  from types import ModuleType
+  from typing import Protocol, TypedDict
+
+
+  class RequiredEditableInfo(TypedDict, total=True):
+      metadata_for_build_editable: str
+      """distribution information of the package as defined by PEP-491"""
```
Comment (jugmac00) [inline pep-9999.rst:145]: I have never seen a docstring for a type annotation - is this common?

Comment (jugmac00) [inline pep-9999.rst:145]: I have never this a docstring for a type annotation... is this common?

File: pep-9999.rst [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+
+  class EditableInfo(RequiredEditableInfo, total=False):
+      sys_paths: list[str]
+      """folders that should be inserted at the start of the sys.path"""
+
+      purelib_paths: list[str]
+      """files or folders that should be available under the purelib"""
```
Comment (jugmac00) [inline pep-9999.rst:152]: I am not familiar with the terms `purelib` and `platlib` - are those terms something worth adding to the "terminology" section? Or are those super familiar for all who will decide on this PEP?

File: pep-9999.rst [hunk 22]
```diff
@@ -0,0 +1,353 @@
...
+          :param name: the module name to import
+          :returns: the module if it can import, otherwise None
+          """
+
+
+Build frontend requirements
+---------------------------
+
+The build frontend is responsible of setting up the environment for the build
+backend to generate the build editable information. It's also responsible of
```
Comment (jugmac00) [inline pep-9999.rst:179]: `the build editable information` is this a common term?

if not...

```suggestion
backend to generate the information which is necessary for an editable build. It's also responsible of
```

File: pep-9999.rst [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+It's the responsibility of the orchestrator to achieve the editable
+functionality. The orchestrator should provide (one or more) mechanisms
+depending on the amount of data the backend provides and communicate directly
+with the user about which to select.
+
+The orchestrator is responsible of making sure the ``.dist-info`` folder is
+available at runtime within the target interpreter for the
+``importlib.metadata`` and ``importlib.resources`` modules.
+
+The orchestrator must ensure that all install requirements (and should the user
```
Comment (jugmac00) [inline pep-9999.rst:195]:
```suggestion
The orchestrator must ensure that all installation requirements (and should the user
```

File: pep-9999.rst [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+require any extra groups) are installed part of the editable build into the
+target interpreter.
+
+The orchestrator must ensure that console or gui entrypoints are generated
```
Comment (jugmac00) [inline pep-9999.rst:199]:
```suggestion
The orchestrator must ensure that console or GUI entrypoints are generated
```

File: pep-9999.rst [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+depending on the amount of data the backend provides and communicate directly
+with the user about which to select.
+
+The orchestrator is responsible of making sure the ``.dist-info`` folder is
+available at runtime within the target interpreter for the
+``importlib.metadata`` and ``importlib.resources`` modules.
+
+The orchestrator must ensure that all install requirements (and should the user
+require any extra groups) are installed part of the editable build into the
+target interpreter.
```
Comment (jugmac00) [inline pep-9999.rst:197]: The above sentence is pretty tough to understand. I suggest do move the contents within in the parenthesis to go at the end and rephrase to `The same applies to ...`.

Also `are installed part of the editable build into the target interpreter.` needs to be rephrased.

File: pep-9999.rst [hunk 26]
```diff
@@ -0,0 +1,353 @@
...
+
+The orchestrator must ensure that console or gui entrypoints are generated
+during the editable installation, just as they would be should the
+``.dist-info`` come as part of a wheel.
```
Comment (jugmac00) [inline pep-9999.rst:201]: This sentence is also hard to understand.

File: pep-9999.rst [hunk 27]
```diff
@@ -0,0 +1,353 @@
...
+(i.e. the directory containing ``pyproject.toml``), and the ``dir_info`` value
+must be ``{'editable': true}``.
+
+The orchestrator must not rely on the ``prepare_metadata_for_build_wheel`` hook
+when installing in editable mode. They must use ``build_editable`` and use the
+``.dist-info`` folder returned by that.
+
+If the import hook is provided by the backend the orchestrator should register
+a hook into the import system at interpreter startup and make sure the hook is called
+before any other import mechanisms are attempted.
```
Comment (jugmac00) [inline pep-9999.rst:224]:
```suggestion
before any other import mechanisms are executed.
```

File: pep-9999.rst [hunk 28]
```diff
@@ -0,0 +1,353 @@
...
+
+The import hook will be called during the import discovery mechanism of the
+target interpreter and there are no guarantees related to the current working
+guarantee or environment variables. The build backend should generate an import
```
Comment (jugmac00) [inline pep-9999.rst:228]: I do not understand this sentence.

File: pep-9999.rst [hunk 29]
```diff
@@ -0,0 +1,353 @@
...
+hook that does not really on the value of these states.
```
Comment (jugmac00) [inline pep-9999.rst:229]:
```suggestion
hook that does not rely on the value of these states.
```

File: pep-9999.rst [hunk 30]
```diff
@@ -0,0 +1,353 @@
...
+
+Should the orchestrator be unable to create an editable installation with the
+set of information returned by the build backend it's free to raise an error
+to the user.
+
+Example editable mode implementations
```
Comment (jugmac00) [inline pep-9999.rst:235]: Until now it was always about "editable installation" - and now there is "editable mode". Is this intended? What is the difference?

File: pep-9999.rst [hunk 31]
```diff
@@ -0,0 +1,353 @@
...
+::
+
+    {
+      "metadata_for_build_editable": "<dir to dist-info>",
+      "sys_paths": ["<project dir>"]
+    }
+
+The orchestrator then could generate a ``pth`` file to insert ``<project dir>``
+onto the ``sys.path`` at startup, and similarly with the parent of the
+distribution information.
```
Comment (jugmac00) [inline pep-9999.rst:265]: I do not understand the last part of the sentence. What distribution information? What is a parent of an information?

File: pep-9999.rst [hunk 32]
```diff
@@ -0,0 +1,353 @@
...
+
+Symlink the source code
+~~~~~~~~~~~~~~~~~~~~~~~
+
+This is what flit does via its `flit install --symlink`_. This solution
+requires the current platform to support symlinks, but allows potentially to
+symlink not folders, but also individual files, which could solve the problem
+of exposing files that otherwise would not be exposed.
+
+This mode requires the build backend to provides one or more folder(s)/file(s)
```
Comment (jugmac00) [inline pep-9999.rst:275]:
```suggestion
This mode requires the build backend to provide one or more folder(s)/file(s)
```

File: pep-9999.rst [hunk 33]
```diff
@@ -0,0 +1,353 @@
...
+
+  {
+    "metadata_for_build_editable": "<dir to dist-info>",
+    "purelib_paths": ["<module_file>", "<package_folder>"]
+  }
+
+The orchestrator then could generate symlinks into the ``purelib`` folder of
+the target interpreter for the ``.dist-info`` folder and all file/folders
+within the ``purelib_paths``. Note should the orchestrator wish they could
+use ``pth`` files with this set of information too.
```
Comment (jugmac00) [inline pep-9999.rst:290]:
```suggestion
use ``pth`` files with this set of information, too.
```

**Review by jugmac00 — COMMENTED** [review]
May the packaging gods be with you and the PEP get accepted! :+1:

File: pep-9999.rst (hunk 19 above)
Comment (gaborbernat) [reply pep-9999.rst:123]: In case it has an import-hook the mechanism that generates the import hook might require additional dependency (e.g. the editables pypi project).

File: pep-9999.rst (hunk 20 above)
Comment (gaborbernat) [reply pep-9999.rst:145]: Can't tell, but definitely can be used.

File: pep-9999.rst (hunk 21 above)
Comment (gaborbernat) [reply pep-9999.rst:152]: Generally, packaging people are super familiar with it. And is already covered in the linked PEP-491 - https://www.python.org/dev/peps/pep-0491/#what-s-the-deal-with-purelib-vs-platlib

File: pep-9999.rst [hunk 34]
```diff
@@ -0,0 +1,359 @@
...
+file, and if it is greater than the current C-extension binary trigger an
+update by calling the compiler before import.
+
+Rejected ideas
+==============
+
+This PEP competes with ``PEP-660``, and rejects that proposal because we think
+the mechanism of achieving an editable installation should be within the build
+frontend rather than the build backend. Furthermore, this approach allows for
+the ecosystem to use alternative mechanisms to achieve the editable install
```
Comment (jugmac00) [inline pep-9999.rst:323]:
```suggestion
the ecosystem to use alternative mechanisms to achieve the editable installation
```

File: pep-9999.rst [hunk 35]
```diff
@@ -0,0 +1,359 @@
...
+effect (e.g., insert path on ``sys.path`` or symlinks instead of implying just
+the loose wheel mode from the backend described by that PEP).
```
Comment (jugmac00) [inline pep-9999.rst:325]:
```suggestion
effect (e.g., insert path on ``sys.path`` or symlinks instead of just implying
the loose wheel mode from the backend described by that PEP).
```

**Review by jugmac00 — COMMENTED** [review]
As promised...

File: pep-9999.rst [hunk 36]
```diff
@@ -0,0 +1,359 @@
...
+
+The Mechanism
+=============
+
+This PEP adds two optional hooks to the `PEP-517`_ backend interface. One of
+the hooks is used to specify the build dependencies of an editable install, and
+the other hook returns the necessary information via the build frontend the
+orchestrator needs to create an editable install.
+
+get_requires_for_build_editable
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply pep-9999.rst:112]: This follows the format set by PEP-517.

File: pep-9999.rst [hunk 37]
```diff
@@ -0,0 +1,247 @@
...
+`PEP-517`_ deferred "Editable installs" because this would have delayed further its adoption, and there wasn't an agreement on how editable installs should be achieved. Due to the popularity of the `setuptools`_ and `pip`_ projects, the status quo prevailed, and the backend could achieve editable mode by providing
+a ``setup.py develop`` implementation, which the user could trigger via `pip install -e <project_directory>`_. By defining an editable interface
+between the build backend and frontend, we can eliminate the ``setup.py`` file and their current communication method.
+
+Terminology and goals
+=====================
+
+The editable installation mode implies that the source code of the project being installed is available in a local directory.
+
+Once the project is installed in editable mode, users expect that changes to the project code in the local source tree become effective without the need for a new installation step.
```
Comment (pganssle) [inline pep-9999.rst:45]: Some users may expect this, but it is by no means a guarantee (it cannot be guaranteed in general).

Perhaps it would be better to phrase this as, "Once the project is installed in editable mode, some changes to the project code in the local source tree will become effective without the need for a new installation step. At minimum, changes to the text of non-generated files that existed at the time of the installation should be reflected upon the next import of the package."

File: pep-9999.rst [hunk 38]
```diff
@@ -0,0 +1,247 @@
...
+
+Some kinds of changes, such as adding or modifying entry points or new dependencies, require a new installation step to become effective. These changes are typically made in build backend configuration files (such as ``pyproject.toml``). This requirement is consistent with the general user expectation that such modifications will only become effective after re-installation.
```
Comment (pganssle) [inline pep-9999.rst:47]: We may want to get into this elsewhere instead. This is one class of changes that may or may not be reflected, depending on the choices of the frontend.

File: pep-9999.rst [hunk 39]
```diff
@@ -0,0 +1,247 @@
...
+
+When a project is installed in editable mode, users expect the installation to behave identically as a standard installation. Depending on how a build
```
Comment (pganssle) [inline pep-9999.rst:49]: It may be best to make it clear that this is almost certainly not possible. In fact, I would start this off as pessimistic, like:

"While users expect editable installations to behave identically to standard installations, this may not always be possible, and may in fact be in tension with other user expectations. Build frontends **should** seek to minimize differences between the behavior of editable and standard installations and document known differences."

File: pep-9999.rst [hunk 40]
```diff
@@ -0,0 +1,247 @@
...
+frontend implements this specification, some minor differences may be visible, such as the presence of additional files (compared to a typical installation),
+either in the source tree or the interpreter's installation path. Both the build backend and frontend are encouraged to document such potential differences.
+
+For reference, a non-editable installation works as follows:
+
+1. The **developer** is using a tool, we'll call it here the **frontend**, to drive the project development (e.g., `pip`_). When the user wants to trigger a package build and installation of a project, they'll communicate with the **frontend**.
+2. The frontend uses a **build frontend** to trigger the build of a wheel (e.g., `build`_). The build frontend uses `PEP-517`_ to communicate with the **build backend** (e.g. `setuptools`_) - with the build backend installed into an isolated environment. Once invoked, the backend returns a wheel.
```
Comment (pganssle) [inline pep-9999.rst:56]: Are you making a distinction between frontend and build frontend? I don't see why there needs to be a distinction here.

Also, it is probably worth changing from "isolated environment" to `PEP-518` environment or something of that nature, since PEP 518 does allow for using non-isolated environments.

File: pep-9999.rst [hunk 41]
```diff
@@ -0,0 +1,247 @@
...
+3. The frontend takes the wheel and feeds it to an **installer** (e.g.,`installer`_) to install the wheel into the target Python interpreter.
+
+The Mechanism
+=============
+
+This PEP adds two optional hooks to the `PEP-517`_ backend interface. One of the hooks is used to specify the build dependencies of an editable install. The other hook returns the necessary information via the build frontend the frontend needs to create an editable install.
+
+get_requires_for_build_editable
```
Comment (pganssle) [inline pep-9999.rst:64]: I'm not sure that this hook is necessary. It's why you'd need special editable-only dependencies. I think this may be residue from PEP 660's suggestion of including a special library for doing editable installations.

I probably would leave it out unless someone says they actually need it, but I don't think it's hurting anything, particularly if it's optional.

File: pep-9999.rst [hunk 42]
```diff
@@ -0,0 +1,247 @@
...
+-------------------------------
+
+::
+
+  def get_requires_for_build_editable(config_settings=None):
+      ...
+
+This hook MUST return an additional list of strings containing `PEP-508`_ dependency specifications, above and beyond those specified in the ``pyproject.toml`` file. These dependencies will be installed when calling the ``build_editable`` hook.
```
Comment (pganssle) [inline pep-9999.rst:72]: Does it need to be a list? Perhaps sequence?

File: pep-9999.rst [hunk 43]
```diff
@@ -0,0 +1,247 @@
...
+
+If not defined, the default implementation is equivalent to return ``[]``.
```
Comment (pganssle) [inline pep-9999.rst:74]: s/return/returning.

File: pep-9999.rst [hunk 44]
```diff
@@ -0,0 +1,247 @@
...
+
+  class RequiredEditableInfo(TypedDict, total=True):
+      version: int
+      """protocol version of the editable metadata, this PEP defines version 1"""
+
+      metadata_for_build_editable: str
+      """distribution information of the package as defined by PEP-491"""
+
+  class EditableInfo(RequiredEditableInfo, total=False):
+      sys_paths: list[str]
```
Comment (pganssle) [inline pep-9999.rst:100]: Generally speaking, I don't use concrete containers, and we definitely shouldn't use them here. This should be `Sequence[str]` and the `dict` ones should be `Mapping`s.

File: pep-9999.rst [hunk 45]
```diff
@@ -0,0 +1,247 @@
...
+
+The function returns an object of type ``EditableInfo`` as defined below:
+
+::
+
+  from types import ModuleType
+  from typing import Protocol, TypedDict
+
+
+  class RequiredEditableInfo(TypedDict, total=True):
```
Comment (pganssle) [inline pep-9999.rst:92]: The prose above calls this `EditableInfo` but this is called `RequiredEditableInfo`.

File: pep-9999.rst [hunk 46]
```diff
@@ -0,0 +1,247 @@
...
+
+build_editable
+--------------
+
+::
+
+  def build_editable(config_settings=None):
+      ...
+
+The function returns an object of type ``EditableInfo`` as defined below:
```
Comment (pganssle) [inline pep-9999.rst:84]: It might be worth adding an "Open Item" where instead of returning a structure like this, we build a wheel-like file that simply contains this information in some serialized form.

Another option would be to have two hooks, one that returns this structure and another that builds a wheel-like object.

File: pep-9999.rst [hunk 47]
```diff
@@ -0,0 +1,247 @@
...
+  class CallbackFunc(Protocol):
+      def __call__(self, name) -> ModuleType | None:
+          """
+          :param name: the module name to import
+          :returns: the module if it can import, otherwise None
+          """
+
+The ``purelib`` and ``platlib`` paths map from project source absolute paths to target directory relative paths. We allow backends to change the project layout from the project source directory to what the interpreter will see by using this format.
+
+Build frontend requirements
```
Comment (pganssle) [inline pep-9999.rst:128]: This distinction between "Build frontend" and "frontend" is incredibly confusing and unnecessary. I would merge them.

File: pep-9999.rst (hunk 42 above)
Comment (pganssle) [reply pep-9999.rst:72]: Also, "These dependencies will be installed when calling the `build_editable` hook" doesn't seem right. It should maybe be "The frontend must ensure that these dependencies are available in the build environment in whcih the `build_editable` hook is called".

File: pep-9999.rst [hunk 48]
```diff
@@ -0,0 +1,247 @@
...
+---------------------------
+
+The build frontend is responsible for setting up the environment for the build backend to generate the necessary information for an editable build. It's also responsible for communicating with the backend and receiving the ``EditableInfo`` object. All recommendations from `PEP-517`_ for the build wheel hook applies here too.
+
+frontend requirements
```
Comment (pganssle) [inline pep-9999.rst:133]: This seems to conflate in a bunch of the backend's responsibilities.

Why do you have the front-end generating the `RECORD` and `dist-info` folders?

My thoughts on this are that an editable install goes like this:

1. Frontend creates the isolated build environment and invokes `build_editable`
2. The backend does everything it normally would for building a wheel *but* instead of zipping up all the files to be installed, it records a mapping of the locations of files that _would have been zipped up_ along with their destination in the wheel file.
    
    The final structure could be a file on disk or it could be a dict-like structure where the `EditableInfo` dictionary is one sub-part. For example:

    ```python
    {
      'dist-info': {
        'RECORD': <full-text>,
        'LICENSE': <full-text>, 'Wheel-version: ...'
      },
      'package_contents': <some_editable>
    }
    ```

    The only reason to do it as an in-memory format like this is that, basically, since it contains no files on disk and all the metadata is likely generated _in memory_ anyway, we can probably save some time and disk write cycles by directly communicating the information. Another option might be to just allow returning an open file-like object.

3. The front-end takes this information and installs all the metadata parts of the wheel as normal, and in place of the package it does whatever it's going to do to ensure that *at least* all of the relevant paths are exposed in the right places.

Almost all of these pieces are basically what each tool was already doing, with the exception that the backend now needs to keep track of where stuff came from and the frontend/installer needs to have a mechanism for creating some sort of active link to the original locations rather than simply unpacking the file in question.

File: pep-9999.rst [hunk 49]
```diff
@@ -0,0 +1,247 @@
+PEP: 9999
+Title: Editable installs
+Author: Bernát Gábor <gaborjbernat@gmail.com>
```
Comment (pganssle) [inline pep-9999.rst:3]: I suppose if you'd like I can be a co-author on the PEP.

**Review by pganssle — CHANGES_REQUESTED** [review]
I've only reviewed as far down as my comments go, but I think a lot of the fundamental structural changes are in the parts I did review.

File: pep-9999.rst [hunk 50]
```diff
@@ -0,0 +1,247 @@
...
+      sys_paths: list[str]
+      """folders that should be inserted at the start of the sys.path"""
+
+      purelib_paths: dict[str, str]
+      """files or folders that should be available under the purelib"""
+
+      platlib_paths: dict[str, str]
+      """files or folders that should be available under the platlib"""
+
+      import_callback: ImportCallback | None
```
Comment (pganssle) [inline pep-9999.rst:109]: What is this? This seems very complicated and unnecessary. I think it should be removed.

File: pep-9999.rst (hunk 48 above)
Comment (gaborbernat) [reply pep-9999.rst:133]: > Why do you have the front-end generating the `RECORD` and `dist-info` folders?

The `RECORD` is a list of files that get installed into the target interpreter and serves as information for the frontend as to what files need to be removed during uninstallation. Considering the frontend is the only one knowing this (because it's derived from the dist-info plus the mappings passed on) only the frontend has all the information available to generate this.
